maetio/mobile-payment-example
# Fully functional Map, includes location search, queries based on users location, and queries based on the map.

### 📋 **Initial Requirements**
*Please ensure that you have checked your PR for the following requirements:*  
1. ✅ **Compilation**: You have run `yarn tsc`, checking for no typescript errors. 
2. ✅ **Linting**: You have run `yarn lint` and fixed all linting and formatting errors.
3. ✅ **Testing**: You have rigorously tested your own code.
4. ✅ **Documentation**: If needed, the documentation is updated with any relevant information or links to new resources.  
5. ✅ **Developer Values**: Code is *modular*, *well-communicated*, *responsive*, and as *lightweight* as possible.


### 🔄 **What was changed?**
___  
Added a lot of useful functionality to the app and map feature. This includes:
* Adding the ability to query data based on the map location. This populated the map with markers, as well as making a Flatlist below the map showing off items.(Image down below)

<div id="header" align="center">
<img src="https://user-images.githubusercontent.com/95386742/198341751-6ef86fd4-9721-42ca-a080-d5939ca58f76.jpg" width="200"/>
</div>

* Fixed navigation, so these Flatlist products and map markers could navigate on touch.
* Added google autocomplete API to allow users to search locations around the world and fetch useful information about that location. This also moved the location of the map to the location searched.(Image down below)

<div id="header" align="center">
<img src="https://user-images.githubusercontent.com/95386742/198341763-9fa92391-69b5-4e51-8095-0be32ada5d54.jpg" width="200"/>
</div>

* Fixed all the type errors that had been in the project prior to this PR




### 🤔 **Why were these changes made?**
___  
Prior to this PR, the map feature was able to find the user's location and display it on the map when they loaded that app,  and query all the friebase data and display it on the map, but that was its only functionality. With this PR the map could now dynamically add markers on the map and flatlist accurately based on the location of the map. It also allows users to search locations and have the map change location to that region selected by the input. All of these changes were implemented for a much nicer user experience with the app. This makes the map feature feel much more polished and similar to popular apps such as Airbnb or Uber. The typescript fixes were made to make the code clean and readable for any future developers.


### ⚙️ **How to test:**
___ 
1. Go to the map-view screen on the bottom navigator in the app.
2. Click "Allow to use my location"
3. You should be able to now see a map with your current locations with a marker. Move around the map and check and see if the coordinates on the bottom of the screen are moving properly.
4. Firebase data only contains data in Thailand, so move map to that part of the world. This should add markers to the map as well as make flatlist items below the map.
5. Click on these items and navigate to the product-view-screen to test navigation fixes.
6. Back on the map screen input a location in the search bar. Click the location and see if the map responded to that input and accurately shows the location typed.


### 🔗 **Useful resources/links**:
___  
- Important resource that fixed a performance issue in react-native-maps. [Here](https://github.com/react-native-maps/react-native-maps/issues/3863) and [Here](https://github.com/react-native-maps/react-native-maps/issues/3639)  (Fixed the stuttery map when trying to manually relocate the map with google autocomplete)
- [Expo Location](https://docs.expo.dev/versions/latest/sdk/location/) To get the user's current location.
- [Expo Map View](https://docs.expo.dev/versions/latest/sdk/map-view/)
- [React Native Maps](https://github.com/react-native-maps/react-native-maps) Used for displaying the map, and markers. Also gives back lots of important information such as location on the map(in Latitude/Longitude)
- [GeoFire](https://firebase.google.com/docs/firestore/solutions/geoqueries) used for querying data based on location.
- [Google Auto-Complete](https://developers.google.com/maps/documentation/javascript/place-autocomplete) and using it react-native [Here](https://github.com/FaridSafi/react-native-google-places-autocomplete) Gives back Lat/Long of cities around the world based on search.
